Fix `__oe_get_enclave_size()` to return ELRANGE size

CodeMonkeyLeet opened the pull request:
`__oe_get_enclave_size()` is used in the `oe_is_within_enclave()` and `oe_is_outside_enclave()` checks, but currently only reflects the initially committed enclave pages during enclave creation, not the actual linear address range (ELRANGE) protected by SGX. This can create unexpected failures in SGX1 as a write to a unbacked page in ELRANGE as if it is outside the enclave will fail.

- Update `oe_sgx_build_enclave()` to patch oeimage with `enclave_size` instead of `enclave_end` (size of committed pages only).

- Remove extra EEID logic to provide the incorrect initial committed size as the enclave size. That code path behaves as in the normal case now.

Also updates tests/eeid_plugin to import the required OE system and platform EDLs;
PR #2961 refactored all tests to explicitly import their required EDLs, but did not update the EEID tests because they only build with the WITH_EEID experimental feature flag enabled.

Fixes #3056

CodeMonkeyLeet commented:
@wintersteiger, note that I've made some fairly minor changes and fixups to the EEID stack.

CodeMonkeyLeet commented:
bors try

wintersteiger commented:
@CodeMonkeyLeet Thanks for that, now I understand what you meant! 

I think there's still a problem with this though: previously, `oe_eeid_extended_size` held the _extended_ size, computed by the second call to `_calculate_enclave_size` in `_add_eeid_marker_page`. Note that this uses different `props`. After your changes, `oe_enclave_properties_sgx.image_info.enclave_size` will hold the _base_ image sizes from the first call to `_calculate_enclave_size`, not the extended sizes. (Patching is the same for both, so as not to change the base image hash.) I think you're right in that we want `enclave_size` instead of `enclave_end`, but we still need another way to track the actual enclave size after extension. Since `oe_enclave_properties_sgx` is const, we can't overwrite it in there (I had a mutable version in an earlier draft), which is why I added `oe_eeid_extended_size`.

CodeMonkeyLeet commented:
> I think there's still a problem with this though: previously, `oe_eeid_extended_size` held the _extended_ size, computed by the second call to `_calculate_enclave_size` in `_add_eeid_marker_page`. Note that this uses different `props`. After your changes, `oe_enclave_properties_sgx.image_info.enclave_size` will hold the _base_ image sizes from the first call to `_calculate_enclave_size`, not the extended sizes. (Patching is the same for both, so as not to change the base image hash.) I think you're right in that we want `enclave_size` instead of `enclave_end`, but we still need another way to track the actual enclave size after extension. Since `oe_enclave_properties_sgx` is const, we can't overwrite it in there (I had a mutable version in an earlier draft), which is why I added `oe_eeid_extended_size`.

@wintersteiger thanks for pointing out the semantic inconsistency this introduced into `_add_eeid_marker_page()`, I've cleaned that up. 

To the broader point of whether the `oe_eeid_extended_size` is necessary, it's not actually used anywhere and my inclination is to remove it until it needs to be reintroduced.

In the master branch, it's semantically incorrect anyway because it was being assigned the `enclave_end` value that was calculated _before_ the dynamic size extension in the body of `_add_eeid_marker_page()`. The `enclave_end` value for the EEID extended size should be the offset to the EEID since it's now written after the extended data pages, so there's no need for a separate variable; we could reconstruct it from the existing `oe_eeid` pointer and enclave base address already available in the enclave.

wintersteiger commented:
Thanks for the fixes @CodeMonkeyLeet! 

You're right, we could re-compute the extended size in the enclave, but we would have to do so at every call to `__oe_get_enclave_size`, because we would not be able to store the result anywhere. It's not very expensive though. Further, we would need to copy the implementation of `_calculate_enclave_size` or move it to `common/*` to make it available in the enclave. I figured saving the size on the host and then writing it to a new global was the option with the least amount of downsides, but I may of course have been wrong. 

Another somewhat tempting option is to cast the `const` on `oe_enclave_properties_sgx` away for just one very brief moment... :-)

CodeMonkeyLeet commented:
@wintersteiger Yes, I agree that there's a slight arithmetic cost to deriving the extended base enclave size from existing variables. We could cache the value in the future if it turns out to be a performance hit, but nothing uses the value today, so I don't think it should be implemented until needed.

My intuition is that this discussion will be relevant again once OE SDK implements support for EDMM. Independent of EEID, the base OE enclave runtime would need to track the allocated enclave memory boundary, and while that value would start as the base image size, it would no longer be treated as a constant in the enclave.

wintersteiger commented:
@CodeMonkeyLeet That's a good point! Since this will have to be dynamic soon we may as well start making it dynamic now. We do currently use that in SGX-LKL, but only during startup and we cache the value, so at least in my current use-case it won't make a difference at all.

radhikaj commented:
bors r+

radhikaj commented:
bors r-

radhikaj commented:
bors r+

wintersteiger commented:
@CodeMonkeyLeet Just checking: do you want me to add the size calculation? Not sure I can push to your fork though.

CodeMonkeyLeet commented:
@wintersteiger No, thanks for the offer, but we should probably take that as a separate change; as noted previously, nothing currently requires it in the codebase, and I'd like to have a separate conversation with @jhand2 first around the form this dynamic sizing should take.

CodeMonkeyLeet commented:
Merge blocked by infrastructure failures to contact YUM servers. Retrying.

bors r+

wintersteiger commented:
@CodeMonkeyLeet  SGX-LKL uses it, i.e. it will report out-of-memory if we return the base image heap size.

CodeMonkeyLeet commented:
@wintersteiger I'm confused, what's the concern about getting base image heap size unexpectedly? How are you getting the base image heap size through OE today?

> i.e. the `oe_eeid_extended_size` never referred to the extended base enclave size to begin with. This PR just brings the normal codepath in line with the existing behavior of the EEID codepath.

The existing behavior of `__oe_get_enclave_size()` gives you the full virtual enclave size already and this PR doesn't change that.

wintersteiger commented:
In the case of EEID base images, the heap and stack size settings are 0, so 

```
return oe_enclave_properties_sgx.image_info.enclave_size;
```

will return a value that's too small. We need the extended size, which was previously saved in `oe_eeid_extended_size` (and on the marker page).

wintersteiger commented:
Darn, sorry about the noise! I thought that change was in `__oe_get_heap_size`. The one in `__oe_get_enclave_size` is perfectly fine, as you say, nobody uses it :-)
